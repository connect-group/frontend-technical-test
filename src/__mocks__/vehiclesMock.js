const vehiclesMock = [
    {
        id: "xe",
        modelYear: "k17",
        apiUrl: "/api/vehicle_xe.json",
        media: [
            {
                name: "vehicle",
                url: "/images/16x9/xe_k17.jpg",
            },
            {
                name: "vehicle",
                url: "/images/1x1/xe_k17.jpg",
            },
        ],
        detail: {
            id: "xe",
            description:
                "The most advanced, efficient and refined sports saloon that Jaguar has ever produced",
            price: "£30,000",
            meta: {
                passengers: 5,
                drivetrain: ["AWD", "RWD"],
                bodystyles: ["saloon"],
                emissions: {
                    template: "CO2 Emissions $value g/km",
                    value: 99,
                },
            },
        },
    },
    {
        id: "fpace",
        modelYear: "k17",
        apiUrl: "/api/vehicle_fpace.json",
        media: [
            {
                name: "vehicle",
                url: "/images/16x9/fpace_k17.jpg",
            },
            {
                name: "vehicle",
                url: "/images/1x1/fpace_k17.jpg",
            },
        ],
        detail: {
            id: "fpace",
            description: "Jaguar's luxury performance SUV.",
            price: "£40,000",
            meta: {
                passengers: 5,
                drivetrain: ["AWD", "RWD"],
                bodystyles: ["SUV"],
                emissions: {
                    template: "CO2 Emissions $value g/km",
                    value: 100,
                },
            },
        },
    },
    {
        id: "ftype",
        modelYear: "k17",
        apiUrl: "/api/vehicle_ftype.json",
        media: [
            {
                name: "vehicle",
                url: "/images/16x9/ftype_k17.jpg",
            },
            {
                name: "vehicle",
                url: "/images/1x1/ftype_k17.jpg",
            },
        ],
        detail: {
            id: "ftype",
            description: "Pulse-quickening, pure Jaguar sports car.",
            price: "£60,000",
            meta: {
                passengers: 2,
                drivetrain: ["AWD", "RWD"],
                bodystyles: ["COUPÉ", "CONVERTIBLE"],
                emissions: {
                    template: "CO2 Emissions $value g/km",
                    value: 234,
                },
            },
        },
    },
    {
        id: "ipace",
        modelYear: "k21",
        apiUrl: "/api/vehicle_ipace.json",
        media: [
            {
                name: "vehicle",
                url: "/images/16x9/ipace_k21.jpg",
            },
            {
                name: "vehicle",
                url: "/images/1x1/ipace_k21.jpg",
            },
        ],
        detail: {
            id: "ipace",
            description:
                "Introducing Jaguar's first all-electric performance SUV.",
            price: "£65,100",
            meta: {
                passengers: 4,
                drivetrain: ["AWD", "RWD"],
                bodystyles: ["COUPÉ", "CONVERTIBLE"],
                emissions: {
                    template: "CO2 Emissions $value g/km",
                    value: 0,
                },
            },
        },
    },
];
export default vehiclesMock;
