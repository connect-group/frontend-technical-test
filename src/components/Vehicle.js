import React from 'react';
import { config } from '../config';

const Vehicle = ({ vehicleData }) => (
  <section>
    <div className="c-vehicle-list-item-image">
      <img
        src={`${config.server.assetsPrefix}${vehicleData.media[0].url}`}
        alt={`Photo of a Jaguar ${vehicleData.id}`}
      />
    </div>
    <div className="c-vehicle-list-item-info">
      <header>
        <h1 className="c-vehicle-list-item-title">Jaguar {vehicleData.id}</h1>
        <h2 className="c-vehicle-list-item-subtitle">From {vehicleData.price}</h2>
      </header>
      <p
        title={vehicleData.description}
        className="c-vehicle-list-item-desc"
      >
        {vehicleData.description}
      </p>
    </div>
  </section>
);

export { Vehicle };
