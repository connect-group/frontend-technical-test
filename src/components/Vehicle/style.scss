// Mobile
.vehicle {

  &__title {
    font-size: 17px;
    font-weight: bold;
    text-align: center;
    text-transform: uppercase;
  }

  &__price {
    font-size: 16px;
    font-weight: bold;
    text-align: center;
  }

  &__wrapper {
    border-bottom: 2px solid #eef1f1;
    display: grid;
    grid-template-columns: 1fr 4fr;
    position: relative;
    transition: 3s ease-in;
  }

  &__description {
    font-family: "Gill Sans", "Gill Sans MT", Calibri, "Trebuchet MS",
      sans-serif;
    font-size: 15px;
    text-align: center;
  }

  &__content {
    font-size: 14px;
    padding: 9px 18px;
  }

  &__thumbnail--mobile {
    background-repeat: no-repeat;
    background-size: contain;
    padding-top: 100%;
  }

  &__thumbnail--tablet,
  &__thumbnail--desktop {
    background-size: cover;
    padding-top: 56.25%;
  }
}

// Tablet
@media screen and (min-width: 768px) {

  .vehicle {

    &__thumbnail--mobile {
      display: none;
    }

    &__wrapper {
      grid-template-columns: 1fr;
    }

    &__title {
      color: #0c121c;
      margin-top: 10px;

      & > span {
        border-bottom: 2px solid #0c121c;
        border-top: 2px solid #0c121c;
      }
    }

    &__price,
    &__description {
      color: rgb(139, 139, 139);
      font-weight: 200;
    }

    &__title,
    &__price,
    &__description {
      height: 35px;
      text-align: center;
    }
  }
}

@media screen and (max-width: 767px) {

  .vehicle {

    &__thumbnail--tablet,
    &__thumbnail--desktop {
      display: none;
    }
  }
}

@media screen and (min-width: 1024px) {

  .vehicle {

    &__wrapper {
      border: 0;
    }
  }
}
