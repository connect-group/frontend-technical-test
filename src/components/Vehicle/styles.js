import styled from "styled-components";
import { device } from "../../constants";

export const Container = styled.div`
    flex-basis: 100%;
    max-width: 100%;
    box-sizing: border-box;
    text-align: left;
    min-height: auto;
    border: 1px solid #eef1f1;
    display: flex;
    
    .cover{
        width: 50vh;
        height: 100%;
        padding-right: 15px;
        
        img {
            width: 100%;
            height: 100%;
        }
    }
    
    .content {
        padding: 10px;
        width: 150vw;
    
        &__name{
            text-transform: uppercase;
            font-size: 25px;
        }
        
        &__text {
            margin-top: 15px;
            color: #555a5f;
        }
    }
    
    @media ${device.tablet} { 
        flex-basis: 50%;
        max-width: 50%;
        min-height: 290px;
        display: block;
        text-align: center;
        
        .cover{
            width: 100%;
            height: auto;
            padding-right: 0;
        }
        
        .content {
            width: auto;
            
             &__name{
                border-top: 1px solid #000;
                border-bottom: 2px solid #000;
            }
        }
    } 
    
    @media ${device.laptop} {
        flex-basis: 25%;
        max-width: 25%;
        min-height: 290px;
    }
`;
