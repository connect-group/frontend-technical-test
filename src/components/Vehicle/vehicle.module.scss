@import "../../styles/mixins";

$borderColor: #b7b7b7;

.vehicle {
    display: flex;
    font-size: 1rem;

    @include md-up {
        flex-direction: column;
    }
}

.image {
    display: flex;
    width: 30%;

    @include md-up {
        width: 100%;
    }

    img {
        width: 100%;
        aspect-ratio: 1/1;

        @include md-up {
            aspect-ratio: 16/9;
        }
    }
}

.details {
    width: 70%;
    display: flex;
    flex-direction: column;
    justify-content: space-evenly;
    padding: 25px;
    border-bottom: 1px solid $borderColor;

    @include md-up {
        width: 100%;
        border-bottom: none;
        height: 100%;
        align-items: center;
        justify-content: space-around;

        &:nth-child(even) {
            border-left: 1px solid $borderColor;
        }
    }
}

.name {
    text-transform: uppercase;
    font-size: 1.1rem;
    font-weight: bolder;

    @include md-up {
        border-width: 2px 0 2px 0;
        border-style: solid;
        padding: 5px;
        margin-bottom: 1.1rem;
    }
}

.price {
    margin-bottom: 1.2rem;
    font-size: 1.0rem;
    color: rgba(0, 0, 0, 0.65);

    @include md-up {
        margin-bottom: 0.5rem;
    }
}

.description {
    color: rgba(0, 0, 0, 0.55);
    font-size: 0.9rem;
}
