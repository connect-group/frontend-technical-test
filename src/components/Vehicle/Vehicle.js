import React from 'react';
import { VehicleContainer } from "./styles";
import { baseURL } from "../../api/config";

const Vehicle = (props) => {
    const {vehicle} = props;
    console.log('vehicle', vehicle);

    return (
        <VehicleContainer>
            <img alt={vehicle.media[0].name} src={vehicle.media[0].url}/>
            <div>
                <span>{vehicle.media[0].name}</span>
                <span>{vehicle.detail.description}</span>
            </div>
        </VehicleContainer>
    );
};

export default Vehicle;
