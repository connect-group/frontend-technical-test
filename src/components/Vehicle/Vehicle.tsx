import React from "react";
import styles from "./Vehicle.module.scss";

const formatter = Intl.NumberFormat("en-GB", {
  currency: "GBP",
  style: "currency",
  maximumFractionDigits: 0,
});

const Vehicle = () => {
  return (
    <div className={styles.container}>
      <picture className={styles.image}>
        <source media="(min-width:768px)" srcSet="/images/16x9/fpace_k17.jpg" />
        <img src="/images/1x1/fpace_k17.jpg" alt={`Foo driving on road.`} />
      </picture>
      <div className={styles.content}>
        <p className={styles.title}>Vehicle Name</p>
        <p className={styles.price}>{`From ${formatter.format(76350)}`}</p>
        <p className={styles.description}>The pinnacle of refined capability</p>
      </div>
    </div>
  );
};

export default Vehicle;
