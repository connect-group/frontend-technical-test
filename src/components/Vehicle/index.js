import React from 'react';
import styles from './vehicle.module.scss';

export default function Vehicle({data}) {
    const {media, id, description, price} = data;

    return (
        <div className={styles.vehicle} data-testid="vehicle">
            <picture className={styles.image} data-testid="vehicle-image">
                <source media="(min-width: 769px)" srcSet={media[0].url} />
                <img src={media[1].url} alt={media[1].name} />
            </picture>

            <div className={styles.details}>
                <div className={styles.name} data-testid="vehicle-name">{id}</div>
                {price && <div className={styles.price} data-testid="vehicle-price">From: {price}</div>}
                {<div className={styles.description} data-testid="vehicle-description">{description}</div>}
            </div>
        </div>
    )
}
