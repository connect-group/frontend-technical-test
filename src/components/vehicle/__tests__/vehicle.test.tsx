import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import Vehicle from "../";
import useVehiclesStore from "../../../store/vehicles-store";
import vehiclesMock from "../../../__mocks__/vehiclesMock";
import { renderHook } from "@testing-library/react-hooks";
import { act } from "react-dom/test-utils";

describe("<Vehicle />", () => {
    let vehicles;
    beforeEach(() => {
        vehicles = vehiclesMock;
        const { result } = renderHook(() => useVehiclesStore());
        act(() => result.current.setVehicles(vehicles));
    });
    it("renders without crashing", () => {
        const { queryByTestId } = render(<Vehicle id={vehicles[0].id} />);
        expect(queryByTestId("vehicle")).toBeDefined();
    });

    it("Should render Vehicle component", async () => {
        const { queryByTestId, container } = render(
            <Vehicle id={vehicles[0].id} />
        );

        expect(queryByTestId("vehicle-title").textContent).toBe(vehicles[0].id);
        expect(queryByTestId("vehicle-price").textContent).toBe(
            `From ${vehicles[0].detail.price}`
        );
        expect(queryByTestId("vehicle-description").textContent).toBe(
            vehicles[0].detail.description
        );

        expect(container).toMatchSnapshot();
    });

    // it('calls selectVehicle when "Enter" key is pressed', () => {
    //     const vehicleId = "12345";
    //     const mockSelectVehicle = jest.fn();
    //     (useVehiclesStore as unknown as jest.Mock).mockReturnValue({
    //         selectVehicle: mockSelectVehicle,
    //     });

    //     render(<Vehicle id={vehicleId} />);
    //     const vehicleElement = screen.getByTestId("vehicle");
    //     fireEvent.keyDown(vehicleElement, { key: "Enter" });
    //     expect(mockSelectVehicle).toHaveBeenCalledWith(vehicleId);
    // });
    // it("calls selectVehicle on enter key press", () => {
    //     const vehicleId = "12345";
    //     const mockSelectVehicle = jest.fn();
    //     (useVehiclesStore as unknown as jest.Mock).mockReturnValue({
    //         selectVehicle: mockSelectVehicle,
    //     });
    //     render(<Vehicle id={vehicles[0].id} />);
    //     const vehicleElement = screen.getByTestId("vehicle");
    //     fireEvent.keyDown(vehicleElement, { key: "Enter" });
    //     expect(mockSelectVehicle).toHaveBeenCalledWith(vehicleId);
    // });

    // it("does not call selectVehicle on other key press", () => {
    //     const vehicleId = "12345";
    //     const mockSelectVehicle = jest.fn();
    //     (useVehiclesStore as unknown as jest.Mock).mockReturnValue({
    //         selectVehicle: mockSelectVehicle,
    //     });
    //     render(<Vehicle id={vehicleId} />);
    //     const vehicleElement = screen.getByTestId("vehicle");
    //     fireEvent.keyDown(vehicleElement, { key: "Escape" });
    //     expect(mockSelectVehicle).not.toHaveBeenCalled();
    // });
});
