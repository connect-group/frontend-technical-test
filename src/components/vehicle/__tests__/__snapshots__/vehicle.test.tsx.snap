// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Vehicle /> Should render Vehicle component 1`] = `
<div>
  <div
    class="vehicle"
    data-testid="vehicle"
    role="button"
    tabindex="0"
  >
    <div
      class="vehicle__image_wrapper"
    >
      <picture
        class="vehicle__image_wrapper-picture"
      >
        <source
          media="(min-width:768px)"
          srcset="/images/16x9/xe_k17.jpg"
        />
        <source
          media="(max-width:767px)"
          srcset="/images/1x1/xe_k17.jpg"
        />
        <img
          alt="vehicle"
          class="vehicle__image_wrapper-img"
          src="/images/16x9/xe_k17.jpg"
        />
      </picture>
      <button
        class="vehicle__show_more"
        tabindex="-1"
        title="Show vehicle extra informations"
        type="button"
      >
        More
      </button>
    </div>
    <div
      class="vehicle__content"
    >
      <h4
        class="vehicle__content-title"
        data-testid="vehicle-title"
      >
        xe
      </h4>
      <p
        class="vehicle__content-price"
        data-testid="vehicle-price"
      >
        From 
        Â£30,000
      </p>
      <p
        class="vehicle__content-description"
        data-testid="vehicle-description"
      >
        The most advanced, efficient and refined sports saloon that Jaguar has ever produced
      </p>
    </div>
  </div>
</div>
`;
